File: sdk/ts/src/attest.ts
// Attest API - daemon /attest/* endpoints
// EAS schema attest 历史 + 创建

import type { SisoulClient } from "./client.js";
import type {
  AttestCreateRequest,
  AttestCreateResponse,
  AttestEntry,
  AttestHistoryResponse,
} from "./types.js";

export class AttestAPI {
  constructor(private readonly client: SisoulClient) {}

  async history(): Promise<AttestEntry[]> {
    const r = await this.client.get<AttestHistoryResponse>("/attest/history");
    return r.history ?? [];
  }

  async create(req: AttestCreateRequest): Promise<AttestCreateResponse> {
    if (!req.schema) throw new Error("attest.create: schema required");
    if (!req.subject_did) throw new Error("attest.create: subject_did required");
    return this.client.post<AttestCreateResponse>("/attest/create", req);
  }

  async bySchema(schema: string): Promise<AttestEntry[]> {
    const all = await this.history();
    return all.filter((e) => e.schema === schema);
  }

  async since(timestampSec: number): Promise<AttestEntry[]> {
    const all = await this.history();
    return all.filter((e) => e.timestamp >= timestampSec);
  }
}

File: sdk/rust/src/attest.rs
//! Attest API.

use crate::client::SisoulClient;
use crate::errors::{Result, SisoulError};
use crate::types::{
    AttestCreateRequest, AttestCreateResponse, AttestEntry, AttestHistoryResponse,
};

pub struct AttestAPI<'a> {
    c: &'a SisoulClient,
}

impl<'a> AttestAPI<'a> {
    pub fn new(c: &'a SisoulClient) -> Self {
        Self { c }
    }

    pub fn history(&self) -> Result<Vec<AttestEntry>> {
        let r: AttestHistoryResponse = self.c.get_json("/attest/history", &[], false)?;
        Ok(r.history)
    }

    pub fn create(&self, req: &AttestCreateRequest) -> Result<AttestCreateResponse> {
        if req.schema.is_empty() {
            return Err(SisoulError::InvalidArgument("attest.create: schema required".into()));
        }
        if req.subject_did.is_empty() {
            return Err(SisoulError::InvalidArgument("attest.create: subject_did required".into()));
        }
        self.c.post_json("/attest/create", req, false)
    }

    pub fn by_schema(&self, schema: &str) -> Result<Vec<AttestEntry>> {
        Ok(self.history()?.into_iter().filter(|e| e.schema == schema).collect())
    }

    pub fn since(&self, timestamp_sec: u64) -> Result<Vec<AttestEntry>> {
        Ok(self.history()?.into_iter().filter(|e| e.timestamp >= timestamp_sec).collect())
    }
}
